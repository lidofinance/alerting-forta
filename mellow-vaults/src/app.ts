import { getEthersProvider, Finding, ethers } from 'forta-agent'
import { ETHProvider } from './clients/eth_provider'
import { FormatterWithEIP1898 } from './clients/eth_formatter'
import { DataRW } from './shared/mutex'
import * as Winston from 'winston'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/HealthChecker.srv'
import { VaultWatcherSrv } from './services/vault-watcher/VaultWatcher.srv'
import { MultisigWatcherSrv } from './services/multisig-watcher/MultisigWatcher.srv'
import { AclChangesSrv } from './services/acl-changes/AclChanges.srv'
import { MELLOW_SYMBIOTIC_ADDRESS, VAULT_LIST, WSTETH_ADDRESS } from 'constants/common'
import {
  SymbioticWstETH__factory,
  Vault,
  Vault__factory,
  VaultConfigurator,
  VaultConfigurator__factory,
  WStETH__factory,
} from './generated'

export type Container = {
  ethClient: ETHProvider
  VaultWatcherSrv: VaultWatcherSrv
  MultisigWatcherSrv: MultisigWatcherSrv
  AclChangesSrv: AclChangesSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static prepareClient = (provider: ethers.providers.JsonRpcProvider) => {
    const VaultContracts: Vault[] = []
    const VaultConfiguratorContracts: VaultConfigurator[] = []
    VAULT_LIST.forEach((vault) => {
      VaultContracts.push(Vault__factory.connect(vault.vault, provider))
      VaultConfiguratorContracts.push(VaultConfigurator__factory.connect(vault.configurator, provider))
    })
    const mellowSymbioticContract = SymbioticWstETH__factory.connect(MELLOW_SYMBIOTIC_ADDRESS, provider)
    const wStETHContract = WStETH__factory.connect(WSTETH_ADDRESS, provider)
    const ethClient = new ETHProvider(
      provider,
      mellowSymbioticContract,
      VaultContracts,
      VaultConfiguratorContracts,
      wStETHContract,
    )
    return ethClient
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()
      ethersProvider.formatter = new FormatterWithEIP1898()

      const ethClient = App.prepareClient(ethersProvider)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const vaultWatcherSrv = new VaultWatcherSrv(logger, ethClient)
      const multisigWatcherSrv = new MultisigWatcherSrv(logger)
      const aclChangesSrv = new AclChangesSrv(logger, ethClient)

      App.instance = {
        ethClient: ethClient,
        VaultWatcherSrv: vaultWatcherSrv,
        MultisigWatcherSrv: multisigWatcherSrv,
        AclChangesSrv: aclChangesSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
