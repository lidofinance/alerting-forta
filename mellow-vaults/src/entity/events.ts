import { FindingSeverity, FindingType, Log } from 'forta-agent'

export type EventOfNotice = {
  name: string
  address: string
  event: string
  alertId: string
  description: CallableFunction
  severity: FindingSeverity
  type?: FindingType
  uniqueKey?: string
}

export type TransactionDtoTx = {
  data: string
  hash: string
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
  transaction: TransactionDtoTx
}

export type BlockDto = {
  number: number
  timestamp: number
  parentHash: string
}
