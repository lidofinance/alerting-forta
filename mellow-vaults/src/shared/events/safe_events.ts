import { FindingSeverity } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { SafeTX } from 'constants/common'
import { getSafeLink, getSafeTxLink, getTxLink } from '../string'

export const GNOSIS_SAFE_EVENTS_OF_NOTICE = [
  {
    event: 'event AddedOwner(address owner)',
    alertId: 'SAFE-OWNER-ADDED',
    name: '🚨 Gnosis Safe: Owner added',
    description: (safeTx: SafeTX, args: Result) => `New owner ${args.owner} was added to ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event RemovedOwner(address owner)',
    alertId: 'SAFE-OWNER-REMOVED',
    name: '🚨 Gnosis Safe: Owner removed',
    description: (safeTx: SafeTX, args: Result) => `Owner ${args.owner} was removed from ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedFallbackHandler(address handler)',
    alertId: 'SAFE-HANDLER-CHANGED',
    name: '🚨 Gnosis Safe: Fallback handler changed',
    description: (safeTx: SafeTX, args: Result) =>
      `Fallback handler for ${getSafeLink(safeTx)} ` + `was changed to ${args.handler}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedGuard(address guard)',
    alertId: 'SAFE-GUARD-CHANGED',
    name: '🚨 Gnosis Safe: Guard changed',
    description: (safeTx: SafeTX, args: Result) => `Guard for ${getSafeLink(safeTx)} was changed to ${args.guard}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedThreshold(uint256 threshold)',
    alertId: 'SAFE-THRESHOLD-CHANGED',
    name: '🚨 Gnosis Safe: Threshold changed',
    description: (safeTx: SafeTX, args: Result) =>
      `Threshold for ${getSafeLink(safeTx)} was changed to ${args.threshold}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event DisabledModule(address module)',
    alertId: 'SAFE-MODULE-DISABLED',
    name: '🚨 Gnosis Safe: Module disabled',
    description: (safeTx: SafeTX, args: Result) => `Module ${args.module} was disabled for ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event EnabledModule(address module)',
    alertId: 'SAFE-MODULE-ENABLED',
    name: '🚨 Gnosis Safe: Module enabled',
    description: (safeTx: SafeTX, args: Result) => `Module ${args.module} was enabled for ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ExecutionFailure(bytes32 txHash, uint256 payment)',
    alertId: 'SAFE-EXECUTION-FAILURE',
    name: '❌ Gnosis Safe: TX Execution failed',
    description: (safeTx: SafeTX, args: Result) =>
      `[TX](${getSafeTxLink(safeTx)}) execution failed for ` +
      `${getSafeLink(safeTx)}\n` +
      `[blockchain explorer](${getTxLink(safeTx)})`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionSuccess(bytes32 txHash, uint256 payment)',
    alertId: 'SAFE-EXECUTION-SUCCESS',
    name: '✅ Gnosis Safe: TX Executed',
    description: (safeTx: SafeTX, args: Result) =>
      `[TX](${getSafeTxLink(safeTx)}) executed by ${getSafeLink(safeTx)}\n` +
      `[blockchain explorer](${getTxLink(safeTx)})`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionFromModuleFailure(address module)',
    alertId: 'SAFE-EXECUTION-FAILURE-FROM-MODULE',
    name: '❌ Gnosis Safe: Execution failed from module',
    description: (safeTx: SafeTX, args: Result) =>
      `TX execution failed for ${getSafeLink(safeTx)} ` + `from module ${args.module}`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionFromModuleSuccess(address module)',
    alertId: 'SAFE-EXECUTION-SUCCESS-FROM-MODULE',
    name: '✅ Gnosis Safe: Execution success from module',
    description: (safeTx: SafeTX, args: Result) =>
      `Execution success for ${getSafeLink(safeTx)} from module ${args.module}`,
    severity: FindingSeverity.Info,
  },
]
