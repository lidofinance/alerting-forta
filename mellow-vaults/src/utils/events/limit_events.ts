import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'

export const MELLOW_VAULT_INCREASE_LIMIT_EVENT = 'event IncreaseLimit(uint256 amount)'

export const vaultLimitNotice = {
  IncreaseLimit: {
    event: MELLOW_VAULT_INCREASE_LIMIT_EVENT,
    alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
    name: '⚠️ Vault: Symbiotic limit increased',
    description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ',
  },
}
