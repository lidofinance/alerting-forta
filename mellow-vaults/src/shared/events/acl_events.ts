import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { ACL_ROLES } from 'constants/common'

export function getACLEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = [
    '12c18264-900b-48a9-bdee-b193b61a36c6',
    'd45a13a5-940c-439f-96e2-2cec78f9de24',
    '19ba7243-ff94-45af-aa3d-6b12a7f0b83d',
  ]

  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event:
        'event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole)',
      alertId: 'MELLOW-VAULT-ROLE-ADMIN-CHANGED',
      name: 'ðŸš¨ Vault: Role Admin changed',
      description: (args: Result) =>
        `Role Admin for role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was changed from ${args.previousAdminRole} to ${args.newAdminRole} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-GRANTED',
      name: 'ðŸš¨ Vault: Role granted',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was granted to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[1],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-REVOKED',
      name: 'ðŸš¨  Vault: Role revoked',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was revoked to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[2],
    },
  ]
}
