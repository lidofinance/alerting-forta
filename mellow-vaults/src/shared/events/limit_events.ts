import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { MELLOW_SYMBIOTIC_ADDRESS } from 'constants/common'

export function getLimitEvents(): EventOfNotice[] {
  const uniqueKeys = ['27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ']

  return [
    {
      address: MELLOW_SYMBIOTIC_ADDRESS,
      event: 'event IncreaseLimit(uint256 amount)',
      alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
      name: '⚠️ Vault: Symbiotic limit increased',
      description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
