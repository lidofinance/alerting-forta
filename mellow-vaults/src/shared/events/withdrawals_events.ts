import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { VAULT_LIST } from 'constants/common'

export function getProcessAllEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['9d846c64-6858-47c4-afd2-3f3d31da079b']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-WITHDRAWAL-ALL',
      name: 'ℹ️ Vault: Withdrawal all',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - withdrawal, bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}

export function getProcessWithdrawalsEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['e27d8c35-bce0-4901-a3d6-ca232cf68ede']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-PARTIAL-WITHDRAWAL',
      name: '⚠️ Vault: Withdrawal partial',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Medium,
      type: FindingType.Suspicious,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
