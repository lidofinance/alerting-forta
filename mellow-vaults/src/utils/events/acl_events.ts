import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { ACL_ROLES, VAULT_LIST } from 'constants/common'

const MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT =
  'event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole)'

const MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT =
  'event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)'

const MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT =
  'event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)'

export const MELLOW_VAULT_STRATEGY_ACL_EVENTS = [
  MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT,
  MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT,
  MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT,
]

const getBondStrategyNameByAddress = (address: string) =>
  VAULT_LIST.find((vault) => vault.defaultBondStrategy === address)

export const aclNotices: Record<string, EventOfNotice> = {
  RoleAdminChanged: {
    event: MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-ADMIN-CHANGED',
    name: 'ðŸš¨ Vault: Role Admin changed',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role Admin for role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was changed from ${args.previousAdminRole} to ${args.newAdminRole} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: '12c18264-900b-48a9-bdee-b193b61a36c6',
  },
  RoleGranted: {
    event: MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-GRANTED',
    name: 'ðŸš¨ Vault: Role granted',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was granted to ${args.account} by ${args.sender} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: 'd45a13a5-940c-439f-96e2-2cec78f9de24',
  },
  RoleRevoked: {
    event: MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-REVOKED',
    name: 'ðŸš¨  Vault: Role revoked',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was revoked to ${args.account} by ${args.sender} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: '19ba7243-ff94-45af-aa3d-6b12a7f0b83d',
  },
}
