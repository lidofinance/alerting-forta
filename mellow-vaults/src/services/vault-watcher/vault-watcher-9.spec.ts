import { BlockDto } from '../../entity/events'
import * as Winston from 'winston'
import { ethers, Finding } from 'forta-agent'

import { VaultWatcherSrv } from './VaultWatcher.srv'
import { App } from '../../app'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'

const TEST_TIMEOUT = 120_000 // ms

describe('VaultWatchers srv functional tests', () => {
  const mainnet = 1
  const ethProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, mainnet)

  const ethClient = App.prepareClient(ethProvider)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new VaultWatcherSrv(logger, ethClient)

  let runBlock: (blockHashOrNumber: string | number, initBlock: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runBlock = async (blockHashOrNumber, initBlock) => {
      if (initBlock > Number(blockHashOrNumber)) {
        throw new Error('Wrong initial block or argument order')
      }
      const initErr = await service.initialize(initBlock)
      if (initErr instanceof Error) {
        throw initErr
      }
      const block = await ethProvider.getBlock(blockHashOrNumber)
      const blockDto: BlockDto = {
        number: block.number,
        timestamp: block.timestamp,
        parentHash: block.parentHash,
      }

      return service.handleBlock(blockDto)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should find limits integrity medium',
    async () => {
      const findings = await runBlock(20527718, 20527717)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
