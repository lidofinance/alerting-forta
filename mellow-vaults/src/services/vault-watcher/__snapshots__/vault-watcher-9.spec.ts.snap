// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaultWatchers srv functional tests should find limits integrity medium 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "VAULT-WSTETH-LIMITS-INTEGRITY-MEDIUM",
    "description": "Mellow Vault [Decentralized Validator Token] (0x5e362eb2c0706bd1d134689ec75176018385430b) - vaultTotalSupply and vaultUnderlyingTvl different",
    "labels": [],
    "metadata": {},
    "name": "⚠️ Vault: vaultTotalSupply and vaultUnderlyingTvl is not the same",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;
